### PR TITLE
[JENKINS-61409] Websockets: Use AbstractByteBufferCommandTransport to transport messages

### PR DESCRIPTION
**Pull request by Vlatombe**
Downstream of https://github.com/jenkinsci/remoting/pull/373

See [JENKINS-61409](https://issues.jenkins-ci.org/browse/JENKINS-61409).

### Proposed changelog entries

* Fix websocket connections crash if message size is greater than 64Kb (Requires a matching `agent.jar` with remoting 4.3 or later)

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [X] JIRA issue is well described
- [ ] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [x] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a JIRA issue should exist and be labeled as `lts-candidate`

### CODE REVIEW

File: core/src/main/java/jenkins/agents/WebSocketAgents.java [hunk 1]
```diff
@@ -128,18 +127,20 @@ protected void opened() {
...
             LOGGER.finest(() -> "reading block of length " + len + " from " + agent);
-            if (offset == 0 && len == payload.length) {
-                receiver.handle(payload);
-            } else {
-                receiver.handle(Arrays.copyOfRange(payload, offset, offset + len));
+            try {
+                transport.receive(ByteBuffer.wrap(payload, offset, len));
+            } catch (IOException e) {
+                error(e);
+            } catch (InterruptedException e) {
```
Comment (jglick) [inline core/src/main/java/jenkins/agents/WebSocketAgents.java]:
```suggestion
            } catch (IOException | InterruptedException e) {
```

File: core/src/main/java/jenkins/agents/WebSocketAgents.java [hunk 2]
```diff
@@ -149,22 +150,12 @@ protected void error(Throwable cause) {
...
+        class Transport extends AbstractByteBufferCommandTransport {
 
             @Override
-            public void setup(AbstractByteArrayCommandTransport.ByteArrayReceiver bar) {
-                receiver = bar;
-            }
-
-            @Override
-            public void writeBlock(Channel chnl, byte[] bytes) throws IOException {
-                LOGGER.finest(() -> "writing block of length " + bytes.length + " to " + agent);
```
Comment (jglick) [inline core/src/main/java/jenkins/agents/WebSocketAgents.java]: Can we keep logging?

File: core/src/main/java/jenkins/agents/WebSocketAgents.java [hunk 3]
```diff
@@ -149,22 +150,12 @@ protected void error(Throwable cause) {
...
-                try {
-                    sendBinary(ByteBuffer.wrap(bytes)).get();
-                } catch (Exception x) {
-                    x.printStackTrace();
-                    throw new IOException(x);
-                }
+            protected void write(ByteBuffer header, ByteBuffer data) throws IOException {
+                sendBinary(header);
+                sendBinary(data);
```
Comment (jglick) [inline core/src/main/java/jenkins/agents/WebSocketAgents.java]:
```suggestion
                sendBinary(header, false);
                sendBinary(data, true);
```

if `WebSocketSession` is extended in the obvious way

File: test/src/test/java/jenkins/agents/WebSocketAgentsTest.java [hunk 4]
```diff
@@ -99,9 +100,10 @@ public void smokes() throws Exception {
             ).stdout(System.out).start());
             r.waitOnline(s);
             assertEquals("response", s.getChannel().call(new DummyTask()));
+            assertNotNull(s.getChannel().call(new FatTask()));
             FreeStyleProject p = r.createFreeStyleProject();
             p.setAssignedNode(s);
-            p.getBuildersList().add(Functions.isWindows() ? new BatchFile("echo hello") : new Shell("echo hello"));
+            p.getBuildersList().add(Functions.isWindows() ? new BatchFile("echo hello") : new Shell("dd if=/dev/random of=file.txt count=1024 bs=200;cat file.txt"));
```
Comment (jglick) [inline test/src/test/java/jenkins/agents/WebSocketAgentsTest.java]:
```suggestion
            p.getBuildersList().add(Functions.isWindows() ? new BatchFile("echo hello") : new Shell("dd if=/dev/random count=1024 bs=200"));
```

File: test/src/test/java/jenkins/agents/WebSocketAgentsTest.java [hunk 5]
```diff
@@ -122,4 +124,18 @@ public String call() {
         }
     }
 
+    private static class FatTask extends SlaveToMasterCallable<String, RuntimeException> {
```
Comment (jglick) [inline test/src/test/java/jenkins/agents/WebSocketAgentsTest.java]: This should suffice to reproduce on its own (both directions), without the `dd`, right?

**Review by jglick — APPROVED** [review]
Looks right to me. Changelog entry should clarify that you need to update Remoting to match (the protocol is not compatible).

**Comment by jeffret-b** [comment]
Do we need an upgrade guide entry for this or should a revised changelog message be enough?

**Review by jeffret-b — APPROVED** [review]
Looks good. I haven't done any manual testing.

**Comment by jglick** [comment]
> I haven't done any manual testing.

Nor I.

**Comment by daniel-beck** [comment]
> Changelog entry should clarify that you need to update Remoting to match (the protocol is not compatible).

Does this need a minimum remoting version update?

File: test/src/test/java/jenkins/agents/WebSocketAgentsTest.java (hunk 5 above)
Comment (Vlatombe) [reply test/src/test/java/jenkins/agents/WebSocketAgentsTest.java]: Yes, though it doesn't seem to exercize the same code path, as the `dd` was working even before I did any change.

**Comment by Vlatombe** [comment]
> Does this need a minimum remoting version update?

Yes this needs a corresponding change in remoting (https://github.com/jenkinsci/remoting/pull/373)

**Comment by daniel-beck** [comment]
> > Does this need a minimum remoting version update?
> 
> Yes this needs a corresponding change in remoting ([jenkinsci/remoting#373](https://github.com/jenkinsci/remoting/pull/373))

I don't see how that will affect https://github.com/jenkinsci/jenkins/blob/master/core/src/main/java/jenkins/slaves/RemotingVersionInfo.java

**Comment by Vlatombe** [comment]
Only if you use websockets, which is beta.

File: test/src/test/java/jenkins/agents/WebSocketAgentsTest.java (hunk 5 above)
Comment (jglick) [reply test/src/test/java/jenkins/agents/WebSocketAgentsTest.java]: Possibly due to line-oriented output? Try excluding newline characters using `tr` or whatever.

Comment (Vlatombe) [reply test/src/test/java/jenkins/agents/WebSocketAgentsTest.java]: For freestyle, output seems to be buffered in 8k blocks, due to usage of https://github.com/jenkinsci/jenkins/blob/a21f83fb2547a29ee46606405aac9c2624b850a3/core/src/main/java/hudson/util/StreamCopyThread.java#L58

Comment (jglick) [reply test/src/test/java/jenkins/agents/WebSocketAgentsTest.java]: Ah OK. So some remote operations involving lots of data would be affected, others would not, and there is no way to tell which other than reading the code or inserting debug traces.

File: core/src/main/java/jenkins/websocket/WebSocketSession.java [hunk 6]
```diff
@@ -130,6 +131,15 @@ protected void text(String message) {
         }
     }
 
+    @SuppressWarnings("unchecked")
+    protected final void sendBinary(ByteBuffer partialByte, boolean isLast) {
+        try {
+            remoteEndpoint.getClass().getMethod("sendPartialBytes", ByteBuffer.class, boolean.class).invoke(remoteEndpoint, partialByte, isLast);
```
Comment (jglick) [inline core/src/main/java/jenkins/websocket/WebSocketSession.java]: Note that unlike the existing overload, this is blocking. Maybe OK in this context.

**Review by jglick — APPROVED** [review]
By the way I just noticed `BatchMode` in the API. It does not help us, does it?

**Comment by jeffret-b** [comment]
> Does this need a minimum remoting version update?

I agree with the previous statements that we don't need to update the minimum remoting version. I'm sure it's not as clear for the user as it could be, but we have indicated the WebSocket implementation is beta, precisely because we expected we would encounter issues like this. Updating the minimum version could impact many people who aren't using the beta feature.

File: core/src/main/java/jenkins/websocket/WebSocketSession.java (hunk 6 above)
Comment (Vlatombe) [reply core/src/main/java/jenkins/websocket/WebSocketSession.java]: I didn't check behaviour of `BatchMode`.

**Comment by jeffret-b** [comment]
I meant to merge the other one in Remoting, but had this one and merged it. It should be good to go, so it's probably fine to leave it merged. If anyone is concerned, we can revert it.

File: test/src/test/java/jenkins/agents/WebSocketAgentsTest.java [hunk 7]
```diff
@@ -99,9 +100,10 @@ public void smokes() throws Exception {
             ).stdout(System.out).start());
             r.waitOnline(s);
             assertEquals("response", s.getChannel().call(new DummyTask()));
+            assertNotNull(s.getChannel().call(new FatTask()));
             FreeStyleProject p = r.createFreeStyleProject();
             p.setAssignedNode(s);
-            p.getBuildersList().add(Functions.isWindows() ? new BatchFile("echo hello") : new Shell("echo hello"));
+            p.getBuildersList().add(Functions.isWindows() ? new BatchFile("echo hello") : new Shell("dd if=/dev/random count=1024 bs=200"));
```
Comment (daniel-beck) [inline test/src/test/java/jenkins/agents/WebSocketAgentsTest.java:106]: Probably safer to use `/dev/urandom` here if you're low on entropy?

Comment (Vlatombe) [reply test/src/test/java/jenkins/agents/WebSocketAgentsTest.java:106]: This line has been reverted by #4605 anyway

Comment (jglick) [reply test/src/test/java/jenkins/agents/WebSocketAgentsTest.java:106]: ⇒ https://github.com/jenkinsci/jenkins/pull/4710#discussion_r420079298

**Pull request merged**